File: circular_queue.cpp
#include <stdio.h>
#include <iostream>

using namespace std;

struct CircularQueue {
  private:
    unsigned capacity;
    unsigned int front_index;
    int* items;
    unsigned size = 0;
    unsigned int tail_index;

    bool is_last(int i) {
      return i == this->capacity - 1;
    }

  public:
    CircularQueue(unsigned int c): capacity(c) {
      this->items = new int[c];
    }

    void clear() {
      while (this->size != 0) {
        this->front_index = this->is_last(this->front_index) ? 0 : this->front_index + 1;
        this->items[this->front_index] = NULL;
        this->size--;
      }
    }

    void dequeue() {
      if (this->empty()) {
        throw runtime_error("[CircularQueue]: is empty!");
      }

      this->front_index = this->is_last(this->front_index) ? 0 : this->front_index + 1;
      this->size--;
    }

    bool empty() {
      return this->size == 0;
    }

    void enqueue(int item) {
      if (this->full()) {
        throw runtime_error("[CircularQueue]: is full!");
      }

      this->tail_index = this->is_last(this->tail_index) ? 0 : this->tail_index--;
      this->items[this->tail_index] = item;
      this->size++;
    }

    int front() {
      return this->items[this->front_index];
    }

    bool full() {
      return this->size == this->capacity;
    }

    void show() {
      int i = this->front_index;

      while (i != this->tail_index) {
        cout << "Item: " << this->items[i] << " | Position: " << i << endl;

        i = this->is_last(i) ? 0 : i + 1;
      }
    }

    int size() {
      return this->size;
    }

    int tail() {
      return this->items[this->tail_index];
    }
};

int main() {
  return 0;
}

File: go/circular_queue/circular_queue.go
package circularqueue

import (
	"errors"
	"fmt"
)

type CircularQueue struct {
	capacity,
	frontIndex,
	tailIndex,
	size uint64
	items []*int64
}

func NewCircularQueue(capacity uint64) *CircularQueue {
	return &CircularQueue{
		capacity:   capacity,
		frontIndex: 0,
		tailIndex:  0,
		size:       0,
		items:      []*int64{},
	}
}

func (cq *CircularQueue) clear() {
	i := cq.frontIndex

	for i != cq.tailIndex {
		cq.items[i] = nil
		i = cq.nextIndex(i)
	}
}

func (cq *CircularQueue) dequeue() {
	if cq.empty() {
		return
	}

	cq.frontIndex = cq.nextIndex(cq.frontIndex)
	cq.size--
}

func (cq *CircularQueue) enqueue(value *int64) error {
	if cq.full() {
		return errors.New("[circular-queue] is full")
	}

	cq.tailIndex = cq.nextIndex(cq.tailIndex)
	cq.items[cq.tailIndex] = value
	cq.size++

	return nil
}

func (cq *CircularQueue) front() *int64 {
	return cq.items[cq.frontIndex]
}

func (cq *CircularQueue) tail() *int64 {
	return cq.items[cq.tailIndex]
}

func (cq *CircularQueue) show() {
	i := cq.frontIndex

	for i != cq.tailIndex {
		fmt.Println(cq.items[i])
		i = cq.nextIndex(i)
	}
}

func (cq *CircularQueue) nextIndex(index uint64) uint64 {
	if index == cq.capacity-1 {
		return 0
	}

	return index + 1
}

func (cq *CircularQueue) full() bool {
	return cq.size == cq.capacity
}

func (cq *CircularQueue) empty() bool {
	return cq.size == 0
}
